### PR TITLE
skipOver should have a single argument with pred version

### PR DESCRIPTION
**Pull request by JackStouffer**
Ping @CyberShadow

### CODE REVIEW

**Comment by CyberShadow** [comment]
Hmm, I guess it is a bit weird that this will skip things in a loop. Meaning, `s.skipOver(" ")` and `s.skipOver!isWhite` are going to do different things.

If we decide to keep the inconsistency, then perhaps returning `size_t` would be more fitting.

This function as it is certainly seems more useful, though. What do you think of naming it `skipWhile`? If we decide we will want a `skipOver` with an alias that skips at most one element, `skipWhile` could be implemented on top of it.

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -4027,6 +4027,64 @@ if (is(typeof(binaryFun!pred(r.front, e))) && isInputRange!R)
...
+ * on the first element which `pred(r.front)` equals `false`.
+ *
+ * Params:
+ *     pred = the test for each element
+ *     r = the range to skip over
+ * Returns:
+ *     `true` if elements were skipped over, `false` otherwise.
+ */
+bool skipOver(alias pred, R)(ref R r) if (ifTestable!(typeof(r.front), unaryFun!pred))
+{
```
Comment (wilzbach) [inline std/algorithm/searching.d]: ## A) `countUntil`

```d
import std.algorithm.iteration : filter;
import std.range;
return (&r).refRange.chain.countUntil!(x => !unaryFun!pred(x)) > 0;
```
## B) `find`

```d
import std.range;
return (&r).refRange.enumerate.find!(x => !unaryFun!pred(x.value)).front.index > 0;
```

**Comment by wilzbach** [comment]
> This function as it is certainly seems more useful, though.

Yes, I often use a combination of the "alternatives" I listed.

> What do you think of naming it skipWhile? If we decide we will want a skipOver with an alias that skips at most one element, skipWhile could be implemented on top of it.

N.B.: `skipWhile = until`, `dropWhile = find`

There's a reason why I made this PR: https://github.com/dlang/phobos/pull/4985
Names in Phobos are sometimes confusing and make it hard to find the needed function
See also:

- https://github.com/dlang/phobos/pull/551
- http://forum.dlang.org/post/mailman.2378.1313523921.14074.digitalmars-d@puremagic.com

tl;dr: no new symbols just for negation (unfortunately)

#### [`findSkip`](https://dlang.org/phobos/std_algorithm_searching.html#findSkip)

`findSkip` in contrast to `find` doesn't accept only a `pred` (it requires a needle)
Adding support for just `pred` would be a trivial alias with this extension ;-)

**Comment by CyberShadow** [comment]
Argh! `find` does not use the eponymous template pattern, which means it's not partially instantiatable with just a `pred`. You should be able to do this:

```d
alias skipWhile(alias pred) = find!(not!(unaryFun!pred));
```

I'd say that we fix `find` if possible, and (especially since Andrei was previously explicitly against this), close this.

**Comment by wilzbach** [comment]
> I'd say that we fix find if possible, and (especially since Andrei was previously explicitly against this), close this.

I would use the opportunity to fix `findSkip` as well ;-)
I can make a PR for these two later today.

**Comment by CyberShadow** [comment]
`find`'s case is a bit complicated because of the two overloads.

**Comment by CyberShadow** [comment]
For this to work, `find`'s overloads must be united into a single template block. We can't really put the unittests inside it, and there's no way to split it up, so this will mean putting the unittests for the one-argument and two-argument overloads together.

**Comment by CyberShadow** [comment]
That, or ignore the two-argument version of `find`... though, it would be nice if that worked as well, so you could do e.g. `alias ifind = find!"toLower(a) == toLower(b)";`.

**Comment by CyberShadow** [comment]
Here is a rough draft for a plan:

```d
// Old definitions
static if (false)
{
	/// Two-argument
	Haystack find(alias pred="a==b", Haystack, Needle)(Haystack h, Needle n)
	{
		return h;
	}

	/// One-argument
	Haystack find(alias pred, Haystack)(Haystack h)
	{
		return h;
	}
}

/// New definitions
static if (true)
{
	/// Two-argument (no pred)
	Haystack find(Haystack, Needle)(Haystack h, Needle n)
	{
		return h;
	}

	template find(alias pred)
	{
		/// Two-argument (with pred)
		Haystack find(Haystack, Needle)(Haystack h, Needle n)
		{
			return h;
		}

		/// One-argument (with pred)
		Haystack find(Haystack)(Haystack h)
		{
			return h;
		}
	}

	/// Backwards compatibility for explicit
	/// find!(pred, Haystack, Needle) instantiations
	template find(alias pred, Haystack, Needle)
	{
		alias findPred = find!pred;
		alias find = findPred!(Haystack, Needle);
	}

	/// Backwards compatibility for explicit
	/// find!(pred, Haystack) instantiations
	template find(alias pred, Haystack)
	{
		alias findPred = find!pred;
		alias find = findPred!Haystack;
	}
}

/////////////////////////////////////

unittest
{
	// No pred
	assert(find("a", "b") == "a");

	// With pred
	assert(find!"a==b"("a", "b") == "a");
	assert(find!"a=='x'"("a") == "a");

	// Explicit instantiations (optional backwards compat)
	assert(find!("a==b", string, string)("a", "b") == "a");
	assert(find!("a=='x'", string)("a") == "a");

	// Partial instantiation (desired)
	alias find1 = find!"a=='x'";
	assert(find1("a") == "a");
	
	alias find2 = find!"a==b";
	assert(find2("a", "b") == "a");
}
```

We'll just need to move the tests for the with-predicate two-argument overload together with the with-predicate one-argument overload.

**Comment by wilzbach** [comment]
> Here is a rough draft for a plan:

Sounds like you volunteered to do it yourself ;-)

**Comment by CyberShadow** [comment]
Argh. There's like 8 overloads!

File: std/algorithm/searching.d [hunk 2]
```diff
@@ -4027,6 +4027,64 @@ if (is(typeof(binaryFun!pred(r.front, e))) && isInputRange!R)
...
+    bool ret = false;
```
Comment (BBasile) [inline std/algorithm/searching.d]: Stop creating AST nodes for something that's done automatically (i.e default initialization).

Comment (CyberShadow) [inline std/algorithm/searching.d]: That is bad advice.

I can't find a quote at the moment, but Walter has been always saying that default initialization is not meant to provide useful default values, but to make using un-initialized variables deterministic and, wherever possible, an error. When possible, the default value of a type is an invalid one (NaN for floating-point types and '\xFF' for chars, which is an invalid UTF-8 code unit); the reason why most other types are initialized to zero is that they do not have an "invalid" value.

So, do initialize your variables explicitly and do not rely on their default value.

Comment (BBasile) [inline std/algorithm/searching.d]: What you didn't know is that the author of the PR is also [the person who suggested to add a D-Scanner check for default initialization](https://github.com/dlang-community/D-Scanner/issues/285). My review comment couldn't be more legit.

Comment (CyberShadow) [inline std/algorithm/searching.d]: Not relevant. The pull request is for Phobos, not some personal repository of Jack's.

File: std/algorithm/searching.d [hunk 3]
```diff
@@ -4027,6 +4027,64 @@ if (is(typeof(binaryFun!pred(r.front, e))) && isInputRange!R)
...
+ * from the start of the range that evaluate to `true`. Stops popping elements
+ * on the first element which `pred(r.front)` equals `false`.
+ *
+ * Params:
+ *     pred = the test for each element
+ *     r = the range to skip over
+ * Returns:
+ *     `true` if elements were skipped over, `false` otherwise.
+ */
+bool skipOver(alias pred, R)(ref R r) if (ifTestable!(typeof(r.front), unaryFun!pred))
```
Comment (andralex) [inline std/algorithm/searching.d]: Nice. Please add but use `/// ditto` for the documentation and consolidate with the existing doc.

File: std/algorithm/searching.d (hunk 1 above)
Comment (andralex) [inline std/algorithm/searching.d]: How do you mean that?

Comment (wilzbach) [inline std/algorithm/searching.d]: > How do you mean that?

It was just an FYI on how this functionality currently can be achieved. Of course, both ways are a bit clunky and won't be as fast as this implementation.

Comment (CyberShadow) [inline std/algorithm/searching.d]: @andralex If we could get `find` to be partially instantiatable with just the predicate then that solves a bunch of problems at once, without introducing new symbols. Unfortunately actually doing it is tricky - see comments below.

**Comment by RazvanN7** [comment]
Next time please add "Fix Issue $Number" so that the dlang bot can associate your PR with the bug fix. That way it can be found at a search through PRs. I couldn't find this PR so I opened another one with almost the same fix, as it can be seen above.

I think that this PR (or mine) should be merged and if consensus is reached with find, then we can modify skipOver. 

My implementation avoids extra initializations and assignments by using a do while. I suggest you take a look before merging this.

**Comment by CyberShadow** [comment]
@andralex What do you think of fixing `find` so it's partially intantiatable with just the predicate (see above)? That would make this PR unnecessary or replaceable with a one-line alias.

File: std/algorithm/searching.d (hunk 1 above)
Comment (andralex) [inline std/algorithm/searching.d]: @CyberShadow I think making find instantiable with the predicate would help a lot of cases. That doesn't necesarily compete with skipOver which is useful for parsing.

**Comment by CyberShadow** [comment]
> I think making find instantiable with the predicate would help a lot of cases.

Agreed, but actually doing so without breaking code seems tricky (though still possible). I was hoping you might have some insights.

> That doesn't necesarily compete with skipOver which is useful for parsing.

Oops, right, I was thinking of `skipWhile` above which returns a new range (like `find`) instead of mutating its ref argument (like `skipOver`).

**Comment by wilzbach** [comment]
> Argh. There's like 8 overloads!

Yeah we first need to reduce them to be able to go ahead.
A first step: https://github.com/dlang/phobos/pull/5575 (more to follow)
